core/psh/psh.c

```diff
@@ -972,7 +974,117 @@ static void psh_signalstop(int sig)
 }
 
 
-static int psh_run(void)
+static int psh_login(void)
+{
+	int insz = -1, code;
+	const int maxinsz = 32;
+	char c, indata[maxinsz+1], indatasave[maxinsz+1], shdw[128];
+	struct termios mode;
+
+	tcgetattr(0, &mode);
+	mode.c_lflag |= (ICANON | ECHO);
+	tcsetattr(0, TCSANOW, &mode);
+
+	/* Get login from user */
+	memset(indata, '\0', sizeof(indata));
```

Comment:
multiple unnecessary cleaning and copying from buffer to buffer... why aren't You putting the data into the target buffer explicitly? Please use meaningful variable names (like `username`, `passwd`) instead of non-descriptive generic ones.

Did You read busybox implementation as advised? (it's much more advanced, but it's good to proof-read it to check if You haven't forgotten anything)

Reply:
After moving the credential getting part to subroutine I used your advise and now all data is saved directly to their desired buffer. Its more elegant to just pass the desired buffer to subroutine. Still, i would prefer to clean the plain text password buffer (which now is named `passwd` as advised) as fast as possible (not to store plaintext password for too long). However cleanups are now limited.

Reply:
I'm not sure if clearing invalid (too long) password as fast as possible is really needed. On the other hand - the valid password will be retained on stack after the function exits :).